tools/fork-network/src/cli.rs

```diff
@@ -1,190 +1,534 @@
-use near_primitives::account::AccessKey;
+use near_crypto::PublicKey;
+use near_epoch_manager::{EpochManager, EpochManagerAdapter, EpochManagerHandle};
+use near_mirror::key_mapping::{map_account, map_key};
+use near_o11y::default_subscriber_with_opentelemetry;
+use near_o11y::env_filter::make_env_filter;
+use near_primitives::account::Account;
+use near_primitives::account::{AccessKey, AccessKeyPermission};
+use near_primitives::block::{Block, BlockHeader};
+use near_primitives::hash::CryptoHash;
+use near_primitives::receipt::Receipt;
+use near_primitives::runtime::config::RuntimeConfig;
 use near_primitives::runtime::config_store::RuntimeConfigStore;
+use near_primitives::shard_layout::ShardUId;
+use near_primitives::state::FlatStateValue;
+use near_primitives::state_record::StateRecord;
+use near_primitives::trie_key::col;
+use near_primitives::trie_key::trie_key_parsers::parse_account_id_from_account_key;
+use near_primitives::types::AccountInfo;
+use near_primitives::types::{AccountId, ShardId, StateRoot};
 use near_primitives::version::PROTOCOL_VERSION;
-use near_primitives::{
-    account::Account,
-    borsh::{self},
-    hash::CryptoHash,
-    types::AccountInfo,
+use near_store::db::RocksDB;
+use near_store::flat::store_helper;
+use near_store::flat::FlatStorageStatus;
+use near_store::{
+    checkpoint_hot_storage_and_cleanup_columns, DBCol, Store, TrieDBStorage, TrieStorage, HEAD_KEY,
 };
-use near_store::{flat::FlatStorageStatus, DBCol, Mode, NodeStorage, HEAD_KEY};
-use nearcore::{load_config, NightshadeRuntime, NEAR_BASE};
+use nearcore::{load_config, open_storage, NearConfig, NightshadeRuntime, NEAR_BASE};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use serde::Deserialize;
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::BufReader;
 use std::path::{Path, PathBuf};
+use std::sync::Arc;
 use strum::IntoEnumIterator;
 
 #[derive(clap::Parser)]
+/// TODO: Write docs.
+/// If --reset, then does [`reset`] and nothing else.
+/// Snapshots `~/.near/data` into `~/.near/data/fork-snapshot/data`.
+/// Checks that hash of flat storage head is the same for every shard.
+/// Reads a block corresponding to the flat head.
+/// Reads a list of validators from `--validators`.
+/// Adds validator accounts to the genesis.
+/// Deletes all columns in `~/.near/data` other than FlatState, State, DbVersion.
+/// Makes a backup of the genesis file in `~/.near/genesis.json.backup`.
+/// Creates a new genesis:
+/// * for chain `$original_chain_id + --chain_id_suffix`
+/// * with epoch_length set to `--epoch_length`
+/// * with contents set to the state roots determined by the block corresponding to the flat head.
 pub struct ForkNetworkCommand {
     #[arg(short, long)]
     pub reset: bool,
     #[arg(short, long, default_value = "1000")]
     pub epoch_length: u64,
+    /// Path to the JSON list of the first epoch validators and their public keys relative to `home_dir`.
```

Comment:
could we document the exact type the json would be parsed as?

Reply:
Done